### DISCUSSION TITLE
Questions regarding fltk-rs

### DISCUSSION THREAD

**Discussion by Joe23232** [fltk-rs/fltk-rs#885] [General]
I am looking for the easiest GUI library out there which is the most supported. So I have some questions:

1. How does `fltk-rs` differ to [`iced`](https://github.com/hecrj/iced) in terms of ease of use and memory/CPU/GPU footprint? Would `fltk-rs` have additional drawbacks in regards to performance as it uses C++ as I heard anything that is OOP will have more drawbacks in regards to its performance?
2. I believe `fltk-rs` is retained graphics, am I correct?
3. Is this suitable for Android development?
4. Can I make my own widgets like this for example:

![Desktop](https://user-images.githubusercontent.com/34926497/132968258-7463cb38-4ab7-47c4-9ca4-d527824ba20b.png)

5. I believe `fltk-rs` doesn't currently support wayland, if this is true would it eventually support wayland?
6. Can it produce similar kind of graphics like [`iced`](https://github.com/hecrj/iced#iced)

![image](https://user-images.githubusercontent.com/34926497/132969198-8b9d3b1b-ce5e-445b-b9d7-1fb35d195c4a.png)
7. Can it be used for web development as well, would it generate into html/css/javascript code?

**Comment by MoAlyousef** [comment]
1- FLTK is known to be light on resources:
https://szibele.com/memory-footprint-of-gui-toolkits/
2- yes
3- no
4- yes
5- it does via xwayland. A dedicated wayland backend is in the works:
https://groups.google.com/g/fltkcoredev/c/Z57t6VziQmQ
6- yes
7- no

**Reply by Joe23232** [reply]
@MoAlyousef Hey man thanks for your response.

3 -  Would it work for Android if some android-ndk or android related library was used? Is the support for Android currently in development?
6 - if I wanted to make something more custom looking I beleive I would have to use OpenGL, am I correct? If I did then would OpenGL consume more system resources?
7 - Is web development support in the development right now and/or would it be suitable in the near future?

**Reply by MoAlyousef** [reply]
3- I can get it to build for Android using Android Studio. For some reason the android-ndk crate doesn't work with fltk-rs. I think it's because it assumes winit's loop or something, I haven't looked closely and can't remember now. Also building and deploying for android requires knowledge of the ndk, and anyways it's not streamlined by the Rust toolchain. So if you wanted to get it building and running on Android, you're on your own.

6- Opengl uses extra resources in the form of your gpu, which is acceptable since it does give 2D and 3D hardware acceleration. FLTK uses the native toolkit to do its drawing, which might also use the gpu under the hood, so it's an implementation detail. X11 might use glx if available, wayland uses egl, Cocoa uses metal, GDI+ on windows uses DirectX, so the gpu might already be used and you might not even know it.

7- You can get the FLTK C++ code base compiled to WASM using emscripten with a few changes, see the discussion here:
https://groups.google.com/g/fltkcoredev/c/A0gL4LZ6-bc/m/0-H8LIVhCAAJ
This is currently not supported in fltk-rs, getting Rust and Emscripten to work together is a major hassle.

**Reply by Joe23232** [reply]
@MoAlyousef I see thanks for your help mate.
